src/magic_folder/test/test_local_snapshot.py

```diff
@@ -0,0 +1,261 @@
+
+
+    @given(lists(path_segments().map(lambda p: p.encode("utf-8")), unique=True),
+           lists(binary(), unique=True))
+    def test_add_multiple_files(self, filenames, contents):
+        files = []
+        for (filename, content) in zip(filenames, contents):
+            file = self.magic_path.child(filename)
+            with file.open("w") as f:
+                f.write(content)
+            files.append(file)
+
+        self.snapshot_creator.add_files(*files)
+        self.snapshot_creator.startService()
+        self.assertThat(
+            self.snapshot_creator.stopService(),
+            succeeded(Always())
+        )
+
+        self.assertThat(self.db.snapshots.keys(), HasLength(len(files)))
+        for (file, content) in zip(files, contents):
+            mangled_filename = path2magic(file.asTextMode(encoding="utf-8").path)
+            stored_snapshot = self.db.get_local_snapshot(mangled_filename, self.author)
+            stored_content = stored_snapshot._get_synchronous_content()
+            self.assertThat(stored_content, Equals(content))
+            self.assertThat(stored_snapshot.parents_local, HasLength(0))
+
+
+    @given(lists(path_segments().map(lambda p: p.encode("utf-8")), unique=True),
+           lists(binary(), unique=True))
+    def test_add_directory_with_files(self, filenames, contents):
+        subdir = mktemp(dir=self.magic_path.path)
+        os.mkdir(subdir)
+
+        files = []
+        for (filename, content) in zip(filenames, contents):
+            file = FilePath(subdir).child(filename)
+            with file.open("w") as f:
+                f.write(content)
+            files.append(file)
+
+        self.snapshot_creator.add_files(FilePath(subdir))
+        self.snapshot_creator.startService()
+        self.assertThat(
+            self.snapshot_creator.stopService(),
+            succeeded(Always())
+        )
+
+        self.assertThat(self.db.snapshots.keys(), HasLength(len(files)))
+
+    @given(content=binary())
+    def test_add_non_filepath(self, content):
+        foo = self.magic_path.child("foo")
+        with foo.open("w") as f:
+            f.write(content)
+
+        with ExpectedException(ValueError,
+                               "every argument to upload_files must be a FilePath"):
+            self.snapshot_creator.add_files(foo.path)
+
+    @given(content1=binary(min_size=1),
+           content2=binary(min_size=1),
+           filename=path_segments().map(lambda p: p.encode("utf-8")),
+    )
+    def test_add_a_file_twice(self, filename, content1, content2):
+        foo = self.magic_path.child(filename)
+        with foo.open("w") as f:
+            f.write(content1)
+
+        self.snapshot_creator.add_files(foo)
+        self.snapshot_creator.startService()
+
+        foo_magicname = path2magic(foo.asTextMode('utf-8').path)
+        stored_snapshot1 = self.db.get_local_snapshot(foo_magicname, self.author)
```

Comment:
The implementation of this test feels a little fishy to me, starting around here.  The test essentially requires that `startService` synchronously create the first `foo` snapshot.  It does now but that doesn't seem like an essential part of the interface.  If snapshot creation changes to be asynchronous (as was the initial intent) then this test will break by failing to find a snapshot at this point.  It would be nice if that didn't happen.

The reason it is awkward to test this here is that the public interface of `snapshot_creator` doesn't support learning when snapshots have been created nor does it have an interface to force snapshots to be created right away.

Perhaps that means `snapshot_creator` should have a different interface.  One idea that comes to mind is that the thing which provides `IService` and does the asynchronous job processing (`_process_queue`) should be one unit which can be tested in isolation and the snapshot creating piece should be a different unit which can also be tested in isolation.  The snapshot creating piece then can have (and needs to have) an interface for making work happen so that the `_process_queue` piece can drive it).  Then *all* these tests for snapshot creation logic can skip over all of the service stuff and async queue stuff and just test the snapshot creation.